### PR TITLE
Add initial range bias implementation

### PR DESCRIPTION
**Pull request by jamesmunns**
Initial PR for range bias correction. Usage looks a little like this:

```rust
use dwm1001::ranging::correction::range_bias::{
    correct_range_bias,
    ch_5_prf_64::CORRECTION_FACTORS,
};

let corrected_distance = correct_range_bias(
    CORRECTION_FACTORS,
    (distance_mm / 1000) as u16,
);
```

### CODE REVIEW

File: src/ranging/correction/range_bias/mod.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            return max(0, min(val, <u16>::max_value() as i32)) as u16;
+        }
+    }
+
+    // NOTE: This is sound if and only if the upper bounds of the correction factor cover the
+    // full u16 range. In this case, it is not possible for the loop not to have already returned
+    // by this point
+    unsafe {
+        unreachable_unchecked();
+    }
```
Comment (hannobraun) [inline src/ranging/correction/range_bias/mod.rs]: I don't think this is correct. This code is reachable, as the user can pass arbitrary data.

I'd prefer a `panic!` with a descriptive message, and an explanation in the documentation, under which circumstances the function panics.

Comment (jamesmunns) [reply src/ranging/correction/range_bias/mod.rs]: Yeah, this is the tradeoff, I should have been more clear that I wasn't 100% on this decision.

I'd say we can either:

1. Only provide the given CorrectionFactors, and then we can guarantee this is unreachable. The generated code will be slightly smaller (but incorrect if the user provides bad values
2. Be conservative in our loop, but accept arbitrary correction curves.

I think I might take the middle option, and call my current impl something like `correct_range_bias_unchecked`, and maybe make it an unsafe function. I'll implement a copy that has an explicit panic if it reaches that point.

Open to other suggestions as well.

File: src/ranging/correction/range_bias/mod.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+pub mod ch_5_prf_64;
+pub mod ch_7_prf_16;
+pub mod ch_7_prf_64;
+
+/// Type alias for correction Factors
+pub type CorrectionFactor = [(u16, i16)];
+
+/// Correct a measurement based on range bias
+pub fn correct_range_bias(
+    correction: &'static CorrectionFactor,
```
[the remark this answers is not in the document]
Comment (jamesmunns) [reply src/ranging/correction/range_bias/mod.rs]: This was probably just laziness, so I didn't have to generic-ize the function. I'll update it to remove the `'static` lifetime.

**Comment by jamesmunns** [comment]
Also, do you have any thoughts about consistent units wrt `cm` vs `mm`? I'd say I can also generate a version of the table in mm so the user doesn't have to do the conversion themselves, but that would take a few more bytes to store the correction table (on the order of 50-100 more bytes). The upside of this is saving a div operation, which can be relatively expensive.

On the other hand, the correction curve only is generally accurate to `cm`, so having a correction factor that is "zero extended" to mm might not make sense.

On the other other hand, we have a hardware floating point unit. Should we just be doing all of this as single precision floating point? I don't know the performance difference off the top of my head between fixed point math (incl multiplication and division), and floating point math that is HW accelerated.

File: src/ranging/correction/range_bias/mod.rs (hunk 1 above)
Comment (hannobraun) [reply src/ranging/correction/range_bias/mod.rs]: Or maybe provide a wrapper type for the correction factors:
``` rust
pub struct CorrectionFactors(&'static [...]);
```
Document the constraint, and provide an `unsafe` constructor. Then the function itself doesn't have to deal with any of it.

**Comment by hannobraun** [comment]
> Also, do you have any thoughts about consistent units wrt `cm` vs `mm`? I'd say I can also generate a version of the table in mm so the user doesn't have to do the conversion themselves, but that would take a few more bytes to store the correction table (on the order of 50-100 more bytes). The upside of this is saving a div operation, which can be relatively expensive.

Oh, I completely missed that. For consistency reasons I'd prefer if everything was in `mm`, but I'm fine with not doing this if it helps with consistency.

Ideally, we'd use something like [uom](https://crates.io/crates/uom), but I haven't had a chance to look into that yet.



> On the other other hand, we have a hardware floating point unit. Should we just be doing all of this as single precision floating point? I don't know the performance difference off the top of my head between fixed point math (incl multiplication and division), and floating point math that is HW accelerated.
> 
> EDIT: The Cortex-M4 has hardware fixed point division: https://community.arm.com/processors/b/blog/posts/divide-and-conquer - disregard most of what I said about performance. The unit consistency is probably important for UX reasons still.

Don't forget, this is the DW1000 driver crate, not the DWM1001 BSP. We really can't make many assumption about the hardware this code is going to run on. Although everything you said is right for our own use case, of course.

File: src/ranging/correction/range_bias/mod.rs (hunk 1 above)
Comment (jamesmunns) [reply src/ranging/correction/range_bias/mod.rs]: Good call. I'll look in to this.

**Comment by jamesmunns** [comment]
Good points, all around.

I think my changes will be:

* Provide a wrapper type for conversion curves, with an unsafe constructor for custom curves
* Move the curves to use `mm`s, to trade `text` size for division operators

**Comment by jamesmunns** [comment]
As a note, I'll probably finish this up once the nrf52 stuff is settled.

**Comment by jamesmunns** [comment]
@hannobraun no worries if this isn't in the first release. I don't think I'll have time in the next week to complete this.

Thanks for the heads up!

**Comment by hannobraun** [comment]
@jamesmunns Understood. Thanks for the update!

**Comment by hannobraun** [comment]
I just noticed something while reviewing issues, specifically #55. The correction values that this pull request introduces are valid for the DWM1001 board and won't apply to the DW1000 in general. Other boards using the DW1000 will need to find their own calibration values. #55 has some hints about where to find more information.

At least the correction values need to live in [`dwm1001`](https://github.com/braun-robotics/rust-dwm1001). I haven't thought through whether it makes sense to keep the rest of this pull request here.

Another consideration: Maybe the `dwm1001` crate should apply these correction values automatically.

**Comment by jamesmunns** [comment]
Closing this as it is stale

**Pull request closed without merging**